### DISCUSSION TITLE
Jinja2 support plan

### DISCUSSION THREAD

**Discussion by thibaudcolas** [torchbox/django-pattern-library#180] [Ideas]
It would be nice for us to have a plan for this? We don’t use Jinja much at Torchbox, so we’d need someone to tell us what would be sensible for this package to do.

**Comment by easherma-truth** [comment]
We use Jinja for our site where I work, but I'm also coming from a similar boat as Torchbox. I'm not familiar enough with Jinja to say what it would take to make this library work with it.

**Comment by gone** [comment]
I've put a first pass at this in #206   - I'm motivated to get this merged in so any feedback would be appreciated! 

I have a sample app up at https://github.com/gone/patternlib_jinja  - if there's interest I can host a copy up on a service for people to poke at.
A couple points that are good to look at:

settings: https://github.com/gone/patternlib_jinja/blob/main/sampleapp/config/settings/base.py#L231-L249
pattern contexts that showcase global/template specific: https://github.com/gone/patternlib_jinja/blob/main/sampleapp/util/pattern_contexts.py

The samples I've been testing against: 
A button - this showcases pulling in some data from yaml, from a global context and a pattern specific context
https://github.com/gone/patternlib_jinja/blob/main/sampleapp/templates/components/button.jinja

A extends/includes example
https://github.com/gone/patternlib_jinja/blob/main/sampleapp/templates/components/child.jinja

As I mentioned in the pr - if you target include_test or extends_test the context isn't included.  Right now my thinking is that we can write some jinja extensions that target those tags, and then users will have to include those extensions into their jinja env config. I'm asking on discord if that's possible or for feedback on that approach.

**Reply by CuriousLearner** [reply]
Hi @gone We've fixed the branch with jinja support and it's here: https://github.com/gone/django-pattern-library/pull/6

@thibaudcolas if you'd like we can open up this directly on the torchbox repo. Please let me know.

**Reply by zerolab** [reply]
@CuriousLearner please do

**Reply by CuriousLearner** [reply]
@zerolab Opened it here: https://github.com/torchbox/django-pattern-library/pull/247

**Comment by thibaudcolas** [comment]
It’s live! Thank you @gone, @luord, @edcohen08, @maribedran, @CuriousLearner for pushing this and making it happen!

I have released the work to date in [version 1.5.0](https://github.com/torchbox/django-pattern-library/releases/tag/v1.5.0). Jinja support is marked as experimental to make it clear that there is still a reasonable likelihood there are unexpected support gaps. For example, as far as I can tell there is no way to override Jinja tags / functions / filters. Anyway, I’ve updated the docs to make it clear there is (some) Jinja support, and also the demo site:

- Source: https://github.com/torchbox/django-pattern-library/tree/main/tests
- Static copy with Jinja templates: https://torchbox.github.io/django-pattern-library/demo/pattern/patterns_jinja/components/accordion/accordion.html

---

If anyone wants to take this further, it would be great to have:

- Feedback that the released Jinja support does work, on a real site 😅 I’ve tried it out on another demo project, unfortunately don’t have a real Jinja setup to try this on.
- More unit tests for the Jinja support. As-is it all requires manual testing, and not all aspects of the pattern library are set up for that.
- [Support for overrides of Jinja tags / functions / filters #255](https://github.com/torchbox/django-pattern-library/issues/255). I don’t think this is super complex but definitely requires a good understanding of how Jinja works. 

I think it’s a matter of manipulating the `environment` here: https://github.com/torchbox/django-pattern-library/blob/main/pattern_library/loader_tags.py#L199, to apply the overrides from the YAML in addition to context changes.

**Reply by LucidDan** [reply]
I'm an active Jinja user on multiple django projects. I've started some work on the bits that I need right now, and will submit PRs for that and anything else I come across that I can help with.

Overall, it did "just work" for me, other than an issue with APP_DIRS that I've already submitted a PR for.

Thank you for this package, and for the Jinja support! It has really improved my UI/UX design workflow.
